sdk/lib/ksanitize/ubsan.c

```diff
@@ -0,0 +1,1775 @@
+{
+
+    ASSERT(pData);
+
+    HandleOverflow(false, pData, ulLHS, ulRHS, MINUS_STRING);
+}
+
+void
+__ubsan_handle_sub_overflow_abort(struct COverflowData *pData, unsigned long ulLHS, unsigned long ulRHS)
+{
+
+    ASSERT(pData);
+
+    HandleOverflow(true, pData, ulLHS, ulRHS, MINUS_STRING);
+}
+
+void
+__ubsan_handle_type_mismatch(struct CTypeMismatchData *pData, unsigned long ulPointer)
+{
+
+    ASSERT(pData);
+
+    HandleTypeMismatch(false, &pData->mLocation, pData->mType, pData->mLogAlignment, pData->mTypeCheckKind, ulPointer);
+}
+
+void
+__ubsan_handle_type_mismatch_abort(struct CTypeMismatchData *pData, unsigned long ulPointer)
+{
+
+    ASSERT(pData);
+
+    HandleTypeMismatch(true, &pData->mLocation, pData->mType, pData->mLogAlignment, pData->mTypeCheckKind, ulPointer);
+}
+
+void
+__ubsan_handle_type_mismatch_v1(struct CTypeMismatchData_v1 *pData, unsigned long ulPointer)
+{
+
+    ASSERT(pData);
+
+    HandleTypeMismatch(false, &pData->mLocation, pData->mType, __BIT(pData->mLogAlignment), pData->mTypeCheckKind, ulPointer);
+}
+
+void
+__ubsan_handle_type_mismatch_v1_abort(struct CTypeMismatchData_v1 *pData, unsigned long ulPointer)
+{
+
+    ASSERT(pData);
+
+    HandleTypeMismatch(true, &pData->mLocation, pData->mType, __BIT(pData->mLogAlignment), pData->mTypeCheckKind, ulPointer);
+}
+
+void
+__ubsan_handle_vla_bound_not_positive(struct CVLABoundData *pData, unsigned long ulBound)
+{
+
+    ASSERT(pData);
+
+    HandleVlaBoundNotPositive(false, pData, ulBound);
+}
+
+void
+__ubsan_handle_vla_bound_not_positive_abort(struct CVLABoundData *pData, unsigned long ulBound)
+{
+
+    ASSERT(pData);
+
+    HandleVlaBoundNotPositive(true, pData, ulBound);
+}
+
+void
+__ubsan_handle_implicit_conversion(struct CImplicitConversionData *pData, unsigned long ulFrom, unsigned long ulTo)
+{
+
+    ASSERT(pData);
+
+    HandleImplicitConversion(false, pData, ulFrom, ulTo);
+}
+
+void
+__ubsan_handle_implicit_conversion_abort(struct CImplicitConversionData *pData, unsigned long ulFrom, unsigned long ulTo)
+{
+    ASSERT(pData);
+
+    HandleImplicitConversion(true, pData, ulFrom, ulTo);
+}
+
+void
+__ubsan_get_current_report_data(const char **ppOutIssueKind, const char **ppOutMessage, const char **ppOutFilename, uint32_t *pOutLine, uint32_t *pOutCol, char **ppOutMemoryAddr)
+{
+    /*
+     * Unimplemented.
+     *
+     * The __ubsan_on_report() feature is non trivial to implement in a
+     * shared code between the kernel and userland. It's also opening
+     * new sets of potential problems as we are not expected to slow down
+     * execution of certain kernel subsystems (synchronization issues,
+     * interrupt handling etc).
+     *
+     * A proper solution would need probably a lock-free bounded queue built
+     * with atomic operations with the property of miltiple consumers and
+     * multiple producers. Maintaining and validating such code is not
+     * worth the effort.
```

Comment:
Hey we might have such a thing... `SLIST`s! ;)

Reply:
"miltiple consumers"? :thinking: